cumulus/parachains/runtimes/assets/common/src/erc20_transactor.rs

```diff
@@ -0,0 +1,220 @@
+use xcm_executor::{
+	traits::{ConvertLocation, Error as MatchError, MatchesFungibles, TransactAsset},
+	AssetsInHolding,
+};
+
+type BalanceOf<T> = <<T as pallet_revive::Config>::Currency as Inspect<
+	<T as frame_system::Config>::AccountId,
+>>::Balance;
+
+/// An Asset Transactor that deals with ERC20 tokens.
+pub struct ERC20Transactor<
+	T,
+	Matcher,
+	AccountIdConverter,
+	GasLimit,
+	StorageDepositLimit,
+	AccountId,
+	TransfersCheckingAccount,
+>(
+	PhantomData<(
+		T,
+		Matcher,
+		AccountIdConverter,
+		GasLimit,
+		StorageDepositLimit,
+		AccountId,
+		TransfersCheckingAccount,
+	)>,
+);
+
+impl<
+		AccountId: Eq + Clone,
+		T: pallet_revive::Config<AccountId = AccountId>,
+		AccountIdConverter: ConvertLocation<AccountId>,
+		Matcher: MatchesFungibles<H160, u128>,
+		GasLimit: Get<Weight>,
+		StorageDepositLimit: Get<BalanceOf<T>>,
+		TransfersCheckingAccount: Get<AccountId>,
+	> TransactAsset
+	for ERC20Transactor<
+		T,
+		Matcher,
+		AccountIdConverter,
+		GasLimit,
+		StorageDepositLimit,
+		AccountId,
+		TransfersCheckingAccount,
+	>
+where
+	BalanceOf<T>: Into<U256> + TryFrom<U256>,
+	MomentOf<T>: Into<U256>,
+	T::Hash: frame_support::traits::IsType<H256>,
+{
+	fn can_check_in(_origin: &Location, _what: &Asset, _context: &XcmContext) -> XcmResult {
+		// We don't support teleports.
+		Err(XcmError::Unimplemented)
+	}
+
+	fn check_in(_origin: &Location, _what: &Asset, _context: &XcmContext) {
+		// We don't support teleports.
+	}
+
+	fn can_check_out(_destination: &Location, _what: &Asset, _context: &XcmContext) -> XcmResult {
+		// We don't support teleports.
+		Err(XcmError::Unimplemented)
+	}
+
+	fn check_out(_destination: &Location, _what: &Asset, _context: &XcmContext) {
+		// We don't support teleports.
+	}
+
+	fn withdraw_asset_with_surplus(
+		what: &Asset,
+		who: &Location,
+		_context: Option<&XcmContext>,
+	) -> Result<(AssetsInHolding, Weight), XcmError> {
+		tracing::trace!(
+			target: "xcm::transactor::erc20::withdraw",
+			?what, ?who,
+		);
+		let (asset_id, amount) = Matcher::matches_fungibles(what)?;
+		let who = AccountIdConverter::convert_location(who)
+			.ok_or(MatchError::AccountIdConversionFailed)?;
+		// We need to map the 32 byte checking account to a 20 byte account.
+		let checking_account_eth = T::AddressMapper::to_address(&TransfersCheckingAccount::get());
+		let checking_address = Address::from(Into::<[u8; 20]>::into(checking_account_eth));
+		let gas_limit = GasLimit::get();
+		// To withdraw, we actually transfer to the checking account.
+		// We do this using the solidity ERC20 interface.
+		let data =
+			IERC20::transferCall { to: checking_address, value: EU256::from(amount) }.abi_encode();
+		let ContractResult { result, gas_consumed, storage_deposit, .. } =
+			pallet_revive::Pallet::<T>::bare_call(
+				T::RuntimeOrigin::signed(who.clone()),
+				asset_id,
+				BalanceOf::<T>::zero(),
+				gas_limit,
+				DepositLimit::Balance(StorageDepositLimit::get()),
+				data,
+			);
```

Comment:
You need to wrap this into a storage transaction and roll back changes on error or revert. `bare_call` itself does not spawn one. Same for all other occurrences where you need to have its changes rolled back on failure.

Or is the whole function rolled back anyways when you return `Err`?

Reply:
The transactor is called by the executor [here](https://github.com/paritytech/polkadot-sdk/blob/e1428b03aada0a74e879b7f80ad3c5903924f026/polkadot/xcm/xcm-executor/src/lib.rs#L904), we always call it inside of a transactional context